File: gplume/src/main/java/com/caibowen/gplume/context/AppContext.java
/*******************************************************************************
 * Copyright 2014 Bowen Cai
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ******************************************************************************/
package com.caibowen.gplume.context;

import java.nio.charset.Charset;
import java.util.Calendar;
import java.util.Date;
import java.util.TimeZone;

import com.caibowen.gplume.context.bean.IBeanAssembler;
import com.caibowen.gplume.context.bean.XMLBeanAssembler;
import com.caibowen.gplume.event.Broadcaster;


/**
 * 
 * life circle management
 * 
 * @author BowenCai
 *
 */
public abstract class AppContext {
	
//	public static final boolean	DEBUG_FLAG = true;
	
//-----------------------------------------------------------------------------
//			global properties
	
	public static class defaults {
		
		public static TimeZone timeZone = TimeZone.getTimeZone("GMT");
		
		static Calendar calendar = Calendar.getInstance(timeZone);

		public static final Charset charSet = Charset.forName("UTF-8");
	}

	/**
	 * thread local variables
	 */
	public static final ThreadLocal<Calendar> currentCalendar = new ThreadLocal<Calendar>() {
		@Override
		protected Calendar initialValue() {
			return Calendar.getInstance(defaults.timeZone);
		}
	};
	
//-----------------------------------------------------------------------------
//				3 basic/top-level components	
	/**
	 * config file location, written in web.xml
	 */
	public static final String MANIFEST = "manifest";
	public static final String LOCALE 	= "locale";
	public static final String TIME_ZONE = "timezone";
	
	public static final IBeanAssembler		beanAssembler = XMLBeanAssembler.instance();
	public static final Broadcaster			broadcaster = Broadcaster.instance();
	
	public static Date now() {
		return defaults.calendar.getTime();
	}

}

File: gplume-scala/src/main/scala/com/caibowen/gplume/scala/context/AppContext.scala
package com.caibowen.gplume.scala.context

import java.nio.charset.Charset
import java.util.{Date, Calendar, TimeZone}

import com.caibowen.gplume.context.{XMLAssembler, IBeanAssembler}
import com.caibowen.gplume.event.Broadcaster

/**
* @author BowenCai
* @since  12/12/2014.
*/
object AppContext {

  object defaults {
    var timeZone = TimeZone.getTimeZone("GMT")
    private[context] val calendar = Calendar.getInstance(timeZone)
    val charSet = Charset.forName("UTF-8")
  }

  /**
   * thread local variables
   */
  val currentCalendar: ThreadLocal[Calendar] =
    new ThreadLocal[Calendar] {
      protected override def initialValue: Calendar = {
        return Calendar.getInstance(defaults.timeZone)
      }
    }
  /**
   * config file location, written in web.xml
   */
  val MANIFEST: String = "manifest"
  val LOCALE: String = "locale"
  val TIME_ZONE: String = "timezone"
  private[this] val _beanBuilder = new com.caibowen.gplume.scala.context.BeanBuilder
  val beanAssembler: IBeanAssembler = new XMLAssembler(_beanBuilder)
  _beanBuilder.setAssembler(beanAssembler)

  val broadcaster = new Broadcaster

  def now: Date = defaults.calendar.getTime

}
